helix-core/src/syntax.rs

```diff
@@ -2024,6 +2024,57 @@ impl<I: Iterator<Item = HighlightEvent>> Iterator for Merge<I> {
     }
 }
 
+pub fn pretty_print_tree<W: fmt::Write>(fmt: &mut W, node: Node<'_>) -> fmt::Result {
+    pretty_print_tree_impl(fmt, node, true, None, 0)
+}
+
+fn pretty_print_tree_impl<W: fmt::Write>(
+    fmt: &mut W,
+    node: Node<'_>,
+    is_root: bool,
+    field_name: Option<&str>,
+    depth: usize,
+) -> fmt::Result {
+    fn is_visible(node: Node<'_>) -> bool {
+        node.is_missing()
+            || (node.is_named() && node.language().node_kind_is_visible(node.kind_id()))
+    }
+
+    if is_visible(node) {
+        let indentation_columns = depth * 2;
+        write!(fmt, "{:indentation_columns$}", "")?;
+
+        if let Some(field_name) = field_name {
+            write!(fmt, "{}: ", field_name)?;
+        }
+
+        write!(fmt, "({}", node.kind())?;
+    } else if is_root {
+        write!(fmt, "(\"{}\")", node.kind())?;
+    }
+
+    for child_idx in 0..node.child_count() {
```

Comment:
This function can be simplified a lot if we use `Node::named_children` here instead of iterating through the children by index. We can eliminate passing `is_root` and checking if nodes are visible:

```rs
pub fn pretty_print_tree<W: fmt::Write>(fmt: &mut W, node: Node<'_>) -> fmt::Result {
    pretty_print_tree_impl(fmt, node, None, 0)
}

fn pretty_print_tree_impl<W: fmt::Write>(
    fmt: &mut W,
    node: Node,
    field_name: Option<&str>,
    depth: usize,
) -> fmt::Result {
    let indentation_columns = depth * 2;
    write!(fmt, "{:indentation_columns$}", "")?;

    if let Some(field_name) = field_name {
        write!(fmt, "{}: ", field_name)?;
    }

    write!(fmt, "({}", node.kind())?;

    for child in node.named_children(&mut node.walk()) {
        fmt.write_char('\n')?;

        pretty_print_tree_impl(
            fmt,
            child,
            node.field_name_for_child(child.id() as u32),
            depth + 1,
        )?;
    }

    write!(fmt, ")")?;

    Ok(())
}
```

The tree-sitter docs recommend using `children`/`named_children` with `TreeCursor`s rather than `child`/`named_child` when you expect to iterate through all of the children: https://github.com/tree-sitter/tree-sitter/blob/cabdcfb41d780c760a2d0bea32f743b7d9ac8e4e/lib/binding_rust/lib.rs#L918-L941

Reply:
So I think the reason I didn't go with the named_children approach is because the child's id doesn't not index correctly. I originally did something similar but decided to just plainly iterate after looking at the `to_sexp` c source in tree-sitter.

Here's the output with the suggested fn:
```
// expected
(function_item
  name: (identifier)
  parameters: (parameters)
  body: (block
    (expression_statement
      (macro_invocation
        macro: (identifier)
        (token_tree
          (string_literal))))))
// output
(function_item
  (identifier)
  (parameters)
  (block
    (expression_statement
      (macro_invocation
        (identifier)
        (token_tree
          (string_literal))))))
```
I wonder if using the iter's index will work instead of the node's id; `.iter().enumerate()`. I'll see if that works...

Reply:
Alright, so using this as the inner loop:
```rust
for (i, child) in node.named_children(&mut node.walk()).enumerate() {
     fmt.write_char('\n')?; 
     pretty_print_tree_impl(fmt, child, node.field_name_for_child(i as u32), depth + 1)?;
}
```
Leads to this:
```
// expected:
(function_item
  name: (identifier)
  parameters: (parameters)
  body: (block
    (expression_statement
      (macro_invocation
        macro: (identifier)
        (token_tree
          (string_literal))))))
// printed
(function_item
  (identifier)
  name: (parameters)
  parameters: (block
    (expression_statement
      (macro_invocation
        macro: (identifier)
        (token_tree
          (string_literal))))))
```
I don't know if there's a way to cleanly use the `TreeCursor` here. My original function is a modified version of the c `to_sexp` implementation, which I don't believe uses a cursor either: https://github.com/tree-sitter/tree-sitter/blob/master/lib/src/subtree.c#L894

Reply:
Ah good catch with the field name. In order to use the TreeCursor, I think we would really need lean in to the TreeCursor API: use `TreeCursor::goto_first_child`, `TreeCursor::goto_next_sibling`, `TreeCursor::field_name` etc.

The implementation here I think is more readable than that would be so I think we should just use this 👍